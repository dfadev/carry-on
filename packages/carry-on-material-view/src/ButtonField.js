import React from "react";
import Button from "@material-ui/core/Button";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";
import getFieldStatus from "./getFieldStatus";
import eventHandler from "./eventHandler";

const ButtonField = withStyles(
  {},
  { name: "CoButtonField" }
)(
  ({
    name,
    label = name,
    value: valueProp,
    disabled: disabledProp,
    onClick,
    ...props
  }) => (
    <Field path={name} type="button">
      {(field, store) => {
        const { disabled, value } = getFieldStatus(
          field,
          store,
          valueProp,
          disabledProp
        );

        return (
          <Button
            {...props}
            disabled={disabled}
            {...field.element}
            value={value}
            onClick={eventHandler(
              onClick,
              props.id || field.element.id,
              value,
              store
            )}
          >
            {label}
          </Button>
        );
      }}
    </Field>
  )
);

export default ButtonField;
