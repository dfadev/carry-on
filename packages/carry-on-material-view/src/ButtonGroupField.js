import React from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import Button from "@material-ui/core/Button";
import ButtonGroup from "@material-ui/core/ButtonGroup";
import Grid from "@material-ui/core/Grid";
import { Field } from "carry-on-react-forms";
import getFieldStatus from "./getFieldStatus";
import eventHandler from "./eventHandler";

const ButtonGroupField = withStyles(
  {},
  { name: "CoButtonGroupField" }
)(
  ({
    name,
    label = null,
    value: valueProp,
    disabled: disabledProp,
    buttons
  }) => (
    <Field path={name}>
      {(field, store) => {
        if (!buttons || buttons.length === 0) return null;

        const { disabled, value } = getFieldStatus(
          field,
          store,
          valueProp,
          disabledProp
        );

        let idIndex = 0;

        const buttonGroup = (
          <ButtonGroup value={value} disabled={disabled}>
            {buttons.map(
              ({
                value: val,
                content,
                onClick,
                // eslint-disable-next-line no-plusplus
                id = `${field.element.id}[${idIndex++}]`,
                ...rest
              }) => (
                <Button
                  key={id}
                  id={id}
                  value={val}
                  onClick={eventHandler(onClick, id, val, store)}
                  {...rest}
                >
                  {content}
                </Button>
              )
            )}
          </ButtonGroup>
        );

        if (label !== undefined && label !== null) {
          return (
            <Grid container spacing={1}>
              <Grid item xs={12}>
                {label}
              </Grid>
              <Grid item>{buttonGroup}</Grid>
            </Grid>
          );
        }

        return buttonGroup;
      }}
    </Field>
  )
);

export default ButtonGroupField;
