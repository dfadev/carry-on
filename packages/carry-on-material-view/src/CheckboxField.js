import React from "react";
import Checkbox from "@material-ui/core/Checkbox";
import FormControl from "@material-ui/core/FormControl";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormHelperText from "@material-ui/core/FormHelperText";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";
import getFieldStatus from "./getFieldStatus";

const CheckboxField = withStyles(
  {},
  { name: "CoCheckboxField" }
)(
  ({
    name,
    label = name,
    readOnly,
    required,
    disabled: disabledProp,
    value: valueProp,
    labelPlacement,
    ...props
  }) => (
    <Field path={name} type="checkbox" readOnly={readOnly}>
      {(field, store) => {
        const { element } = field;
        const { error, hasError, disabled, value } = getFieldStatus(
          field,
          store,
          valueProp,
          disabledProp
        );

        return (
          <FormControl error={hasError} required={required}>
            <FormControlLabel
              control={
                <Checkbox
                  {...props}
                  inputProps={{ ...element, checked: value }}
                  checked={value}
                />
              }
              label={label}
              disabled={disabled}
              labelPlacement={labelPlacement}
            />
            {hasError && <FormHelperText>{error}</FormHelperText>}
          </FormControl>
        );
      }}
    </Field>
  )
);

export default CheckboxField;
