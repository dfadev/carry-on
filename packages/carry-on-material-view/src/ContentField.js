import React from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";

const ContentField = withStyles(
  {},
  { name: "CoContentField" }
)(({ name, content }) => (
  <Field path={name}>
    {field =>
      content === undefined && content !== null ? field.element.value : content
    }
  </Field>
));

export default ContentField;
