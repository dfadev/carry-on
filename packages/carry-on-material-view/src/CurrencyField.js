import React from "react";
import InputAdornment from "@material-ui/core/InputAdornment";
import withStyles from "@material-ui/core/styles/withStyles";
import MaskedTextField from "./MaskedTextField";

const numberAccept = /[\d.]+/g;

const parseNumber = string => {
  if (typeof string !== "string") string = string.toString();
  return (string.match(numberAccept) || []).join("");
};

const formatFloatingPointNumber = (value, maxDigits) => {
  const parsed = parseNumber(value.slice(0, 19));
  const [head, tail] = parsed.split(".");
  // Avoid rounding errors at toLocaleString as when user enters 1.239 and maxDigits=2 we
  // must not to convert it to 1.24, it must stay 1.23
  const scaledTail = tail != null ? tail.slice(0, maxDigits) : "";

  const number = Number.parseFloat(`${head}.${scaledTail}`);

  if (Number.isNaN(number)) {
    return "";
  }

  const formatted = number.toLocaleString(undefined, {
    minimumFractionDigits: 0,
    maximumFractionDigits: maxDigits
  });

  if (parsed.includes(".")) {
    const [formattedHead] = formatted.split(".");

    // skip zero at digits position for non fixed floats
    // as at digits 2 for non fixed floats numbers like 1.50 has no sense, just 1.5 allowed
    // but 1.0 has sense as otherwise you will not be able to enter 1.05 for example
    const formattedTail =
      scaledTail !== "" &&
      scaledTail.length < maxDigits &&
      scaledTail[maxDigits] === "0"
        ? scaledTail.slice(0, -1)
        : scaledTail;

    return maxDigits === 0
      ? formattedHead
      : `${formattedHead}.${formattedTail}`;
  }
  return formatted;
};

const formatCurrency = maxDigits => string =>
  formatFloatingPointNumber(string, maxDigits);

const accept = /[\d.$]/g;

const CurrencyField = withStyles(
  {},
  { name: "CoCurrencyField" }
)(({ InputProps, maximumFractionDigits = 2, symbol = "$", ...props }) => (
  <MaskedTextField
    accept={accept}
    format={formatCurrency(maximumFractionDigits)}
    InputProps={{
      startAdornment: (
        <InputAdornment position="start">{symbol}</InputAdornment>
      ),
      ...(InputProps || {})
    }}
    {...(props || {})}
  />
));

export default CurrencyField;
