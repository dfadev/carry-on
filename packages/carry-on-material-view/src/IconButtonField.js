import React from "react";
import IconButton from "@material-ui/core/IconButton";
import withStyles from "@material-ui/core/styles/withStyles";
import { Field } from "carry-on-react-forms";
import getFieldStatus from "./getFieldStatus";
import eventHandler from "./eventHandler";

const IconButtonField = withStyles(
  {},
  { name: "CoIconButtonField" }
)(
  ({
    name,
    label = name,
    value: valueProp,
    disabled: disabledProp,
    onClick,
    ...props
  }) => (
    <Field path={name} type="button">
      {(field, store) => {
        const { disabled, value } = getFieldStatus(
          field,
          store,
          valueProp,
          disabledProp
        );

        return (
          <IconButton
            {...props}
            disabled={disabled}
            id={field.element.id}
            value={value}
            onClick={eventHandler(
              onClick,
              props.id || field.element.id,
              value,
              store
            )}
          >
            {label}
          </IconButton>
        );
      }}
    </Field>
  )
);

export default IconButtonField;
