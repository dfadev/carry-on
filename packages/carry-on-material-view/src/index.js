export { default as ButtonField } from "./ButtonField";
export { default as CheckboxField } from "./CheckboxField";
export { default as CurrencyField } from "./CurrencyField";
export { default as EINField } from "./EINField";
export { default as IconButtonField } from "./IconButtonField";
export { default as MaskedTextField } from "./MaskedTextField";
export { default as PercentageField } from "./PercentageField";
export { default as PhoneField } from "./PhoneField";
export { default as RegisterMaterialView } from "./RegisterMaterialView";
export { default as ResetButtonField } from "./ResetButtonField";
export { default as SSNField } from "./SSNField";
export { default as SubmitButtonField } from "./SubmitButtonField";
export { default as TextField } from "./TextField";
export { default as ZipCodeField } from "./ZipCodeField";
export { default as getFieldStatus } from "./getFieldStatus";
export {
  default as registerComponents,
  materialViewComponents
} from "./registerComponents";
export { default as SelectField } from "./SelectField";
export { default as RadioField } from "./RadioField";
export { default as SwitchField } from "./SwitchField";
export { default as ToggleButtonGroupField } from "./ToggleButtonGroupField";
export { default as ContentField } from "./ContentField";
export { default as InspectorField } from "./InspectorField";
export { default as DateField } from "./DateField";
export { default as DobField } from "./DobField";
export { default as TimeField } from "./TimeField";
export { default as DateTimeField } from "./DateTimeField";
export { default as ListField } from "./ListField";
export { default as DataField } from "./DataField";
export { default as TabsField } from "./TabsField";
export { default as SubmitIconField } from "./SubmitIconField";
export { default as SubmitFabField } from "./SubmitFabField";
export { default as FabField } from "./FabField";
export { default as ImageField } from "./ImageField";
export { default as ButtonGroupField } from "./ButtonGroupField";
