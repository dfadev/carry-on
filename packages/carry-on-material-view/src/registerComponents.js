import React from "react";
import { register } from "carry-on-store";
import { Box, Paper, Grid } from "@mui/material";
import TextField from "./TextField";
import CheckboxField from "./CheckboxField";
import ButtonField from "./ButtonField";
import IconButtonField from "./IconButtonField";
import CurrencyField from "./CurrencyField";
import SSNField from "./SSNField";
import EINField from "./EINField";
import PercentageField from "./PercentageField";
import PhoneField from "./PhoneField";
import ZipCodeField from "./ZipCodeField";
import SubmitButtonField from "./SubmitButtonField";
import ResetButtonField from "./ResetButtonField";
import SelectField from "./SelectField";
import RadioField from "./RadioField";
import SwitchField from "./SwitchField";
import ToggleButtonGroupField from "./ToggleButtonGroupField";
import ContentField from "./ContentField";
import InspectorField from "./InspectorField";
import DateField from "./DateField";
import DobField from "./DobField";
import TimeField from "./TimeField";
import DateTimeField from "./DateTimeField";
import ListField from "./ListField";
import DataField from "./DataField";
import TabsField from "./TabsField";
import SubmitIconField from "./SubmitIconField";
import SubmitFabField from "./SubmitFabField";
import FabField from "./FabField";
import ImageField from "./ImageField";
import ButtonGroupField from "./ButtonGroupField";

function FormViewBox({ children }) {
  return (
    <Paper elevation={3}>
      <Box m={1} p={1}>
        {children}
      </Box>
    </Paper>
  );
}

function SectionBox({ section, children, ...props }) {
  return <Box {...props}>{children}</Box>;
}

function View({ children, ...props }) {
  return (
    <Grid {...props} container>
      {children}
    </Grid>
  );
}

function ViewItem({ children, field, ...props }) {
  return (
    <Grid {...props} item>
      <Box mb={1} p={1}>
        {children}
      </Box>
    </Grid>
  );
}

export const materialViewComponents = {
  components: {
    FormViewBox,
    SectionBox,
    View,
    ViewItem,
    editors: {
      button: ButtonField,
      buttonGroup: ButtonGroupField,
      checkbox: CheckboxField,
      content: ContentField,
      currency: CurrencyField,
      data: DataField,
      date: DateField,
      dateTime: DateTimeField,
      dob: DobField,
      ein: EINField,
      fab: FabField,
      iconButton: IconButtonField,
      image: ImageField,
      inspector: InspectorField,
      list: ListField,
      percentage: PercentageField,
      phone: PhoneField,
      radio: RadioField,
      reset: ResetButtonField,
      select: SelectField,
      ssn: SSNField,
      submit: SubmitButtonField,
      submitFab: SubmitFabField,
      submitIcon: SubmitIconField,
      switch: SwitchField,
      tabs: TabsField,
      text: TextField,
      time: TimeField,
      toggleButtonGroup: ToggleButtonGroupField,
      zipCode: ZipCodeField
    }
  }
};

function registerComponents(...args) {
  register({ state: materialViewComponents }, ...args);
}

export default registerComponents;
