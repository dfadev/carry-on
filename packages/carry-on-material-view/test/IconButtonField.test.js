/**
 * @jest-environment jsdom
 */
import React from "react";
import { initStores } from "carry-on-store";
import { render } from "@testing-library/react";
import { Paper, Box, Grid } from "@material-ui/core";
import PlayArrowIcon from "@material-ui/icons/PlayArrow";
import PauseIcon from "@material-ui/icons/Pause";
import StopIcon from "@material-ui/icons/Stop";
import FastForwardIcon from "@material-ui/icons/FastForward";
import FastRewindIcon from "@material-ui/icons/FastRewind";
import { Store, Register } from "carry-on-react";
import { Form } from "carry-on-react-forms";
import {
  FormView,
  Fields,
  Field,
  Sections,
  Section
} from "carry-on-react-view";
import { materialViewComponents } from "../src/registerComponents";
import IconButtonField from "../src/IconButtonField";

beforeEach(() => {
  initStores();
});

it("should render <IconButtonField> inside a Form", () => {
  expect(
    render(
      <Store id={"iconButtonFieldForm"}>
        <Form>
          <Paper>
            <Box px={3} pt={2} pb={3} mb={2}>
              <Grid container>
                <Grid item>
                  <IconButtonField name="play" label={<PlayArrowIcon />} />
                </Grid>
                <Grid item>
                  <IconButtonField name="pause" label={<PauseIcon />} />
                </Grid>
                <Grid item>
                  <IconButtonField name="stop" label={<StopIcon />} />
                </Grid>
                <Grid item>
                  <IconButtonField
                    name="fastRewind"
                    label={<FastRewindIcon />}
                  />
                </Grid>
                <Grid item>
                  <IconButtonField name="fastFwd" label={<FastForwardIcon />} />
                </Grid>
              </Grid>
            </Box>
          </Paper>
        </Form>
      </Store>
    ).asFragment()
  ).toMatchSnapshot();
});

it("should render <IconButtonField> inside a FormView", () => {
  expect(
    render(
      <Store id={"iconButtonFieldFormView"}>
        <FormView>
          <Register>{materialViewComponents}</Register>
          <Fields>
            <Field name="play" label={<PlayArrowIcon />} editor="iconButton" />
            <Field name="pause" label={<PauseIcon />} editor="iconButton" />
            <Field name="stop" label={<StopIcon />} editor="iconButton" />
            <Field
              name="fastRewind"
              label={<FastRewindIcon />}
              editor="iconButton"
            />
            <Field
              name="fastFwd"
              label={<FastForwardIcon />}
              editor="iconButton"
            />
          </Fields>
          <Sections>
            <Section>
              {["play", "pause", "stop", "fastRewind", "fastFwd"]}
            </Section>
          </Sections>
        </FormView>
      </Store>
    ).asFragment()
  ).toMatchSnapshot();
});
