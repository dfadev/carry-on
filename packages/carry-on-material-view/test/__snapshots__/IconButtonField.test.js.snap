// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <IconButtonField> inside a Form 1`] = `
<DocumentFragment>
  <form
    id="iconButtonFieldForm.form"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              id="iconButtonFieldForm.form.play"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8 5v14l11-7z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              id="iconButtonFieldForm.form.pause"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              id="iconButtonFieldForm.form.stop"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 6h12v12H6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              id="iconButtonFieldForm.form.fastRewind"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              id="iconButtonFieldForm.form.fastFwd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`should render <IconButtonField> inside a FormView 1`] = `
<DocumentFragment>
  <form
    id="iconButtonFieldFormView.form"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <div
          class="MuiBox-root MuiBox-root-3"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiBox-root MuiBox-root-4"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  id="iconButtonFieldFormView.form.play"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8 5v14l11-7z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiBox-root MuiBox-root-5"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  id="iconButtonFieldFormView.form.pause"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 19h4V5H6v14zm8-14v14h4V5h-4z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiBox-root MuiBox-root-6"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  id="iconButtonFieldFormView.form.stop"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M6 6h12v12H6z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiBox-root MuiBox-root-7"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  id="iconButtonFieldFormView.form.fastRewind"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiBox-root MuiBox-root-8"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  id="iconButtonFieldFormView.form.fastFwd"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
